lib/core/src/send_swap.rs

```diff
@@ -0,0 +1,420 @@
+                    .input
+                    .first()
+                    .ok_or_else(|| anyhow!("Found no input for claim tx"))?;
+
+                let script_witness_bytes = input.clone().witness.script_witness;
+                info!("Found Send Swap {id} claim tx witness: {script_witness_bytes:?}");
+                let script_witness = Witness::from(script_witness_bytes);
+
+                let preimage_bytes = script_witness
+                    .nth(1)
+                    .ok_or_else(|| anyhow!("Claim tx witness has no preimage"))?;
+                let preimage = sha256::Hash::from_slice(preimage_bytes)
+                    .map_err(|e| anyhow!("Claim tx witness has invalid preimage: {e}"))?;
+                let preimage_hex = preimage.to_hex();
+                debug!("Found Send Swap {id} claim tx preimage: {preimage_hex}");
+
+                Ok(preimage_hex)
+            }
+        }
+    }
+
+    async fn validate_send_swap_preimage(
+        &self,
+        swap_id: &str,
+        invoice: &str,
+        preimage: &str,
+    ) -> Result<(), PaymentError> {
+        Self::verify_payment_hash(preimage, invoice)?;
+        info!("Preimage is valid for Send Swap {swap_id}");
+        Ok(())
+    }
+
+    async fn refund(&self, swap: &SendSwap) -> Result<String, PaymentError> {
+        let amount_sat = get_invoice_amount!(swap.invoice);
+        let output_address = self.onchain_wallet.next_unused_address().await?.to_string();
+
+        let fee = self
+            .onchain_wallet
+            .build_tx(None, &output_address, amount_sat)
+            .await?
+            .all_fees()
+            .values()
+            .sum();
+        let broadcast_fees_sat = Amount::from_sat(fee);
+
+        let refund_res =
+            self.swapper
+                .refund_send_swap_cooperative(swap, &output_address, broadcast_fees_sat);
+        match refund_res {
+            Ok(res) => Ok(res),
+            Err(e) => {
+                warn!("Cooperative refund failed: {:?}", e);
+                self.refund_non_cooperative(swap, broadcast_fees_sat).await
+            }
+        }
+    }
+
+    async fn refund_non_cooperative(
+        &self,
+        swap: &SendSwap,
+        broadcast_fees_sat: Amount,
+    ) -> Result<String, PaymentError> {
+        info!(
+            "Initiating non-cooperative refund for Send Swap {}",
+            &swap.id
+        );
+
+        let current_height = self.onchain_wallet.tip().await.height();
+        let output_address = self.onchain_wallet.next_unused_address().await?.to_string();
+        let refund_tx_id = self.swapper.refund_send_swap_non_cooperative(
+            swap,
+            broadcast_fees_sat,
+            &output_address,
+            current_height,
+        )?;
+
+        info!(
+            "Successfully broadcast non-cooperative refund for Send Swap {}, tx: {}",
+            swap.id, refund_tx_id
+        );
+        Ok(refund_tx_id)
+    }
+
+    fn validate_state_transition(
```

Comment:
I notice the `validate_state_transition` function is copied (not moved) from the sdk module. I assume you will move it for the receive swap handler. I guess it's the best way to do it even though at the moment the validation is the same for both send and receive.

Reply:
Next on my plate is to extract out the receive swap handler. If this logic needs to be common I will refactor it out to a shared component.
I just not sure atm that the state transition validation should be the same for both. For example tonedout doesn't exist in receive.